include/exadg/grid/grid.cpp

```diff
@@ -160,10 +160,47 @@ Grid<dim>::create_triangulation(
     {
       // we need to generate the coarse triangulations explicitly
 
-      // TODO
-      // construct all the coarser triangulations
+      // resize the empty coarse triangulations vector
+      coarse_triangulations = std::vector<std::shared_ptr<dealii::Triangulation<dim> const>>(
+        triangulation->n_global_levels());
+
+      unsigned int coarse_level = triangulation->n_global_levels() - 1;
+
+      // lambda function for creating and placing the coarse triangulations
+      auto const lambda_create_coarse_levels = [&](unsigned int              global_refinement,
+                                                   std::vector<unsigned int> refine_local) {
+        auto new_tria = std::make_shared<dealii::parallel::fullydistributed::Triangulation<dim>>(
+          triangulation->get_communicator());
+
+        do_create_triangulation(
+          new_tria, data, create_coarse_triangulation, global_refinement, refine_local);
+
+        coarse_triangulations[coarse_level] = new_tria;
+
+        coarse_level--;
+      };
 
-      AssertThrow(false, ExcNotImplemented());
+      // start with the finest level
+      std::vector<unsigned int> refine_local = vector_local_refinements;
+
+      // undo global refinements
+      for(int it = global_refinements; it >= 0; --it)
+      {
+        lambda_create_coarse_levels(it, refine_local);
+      }
+
+      // undo local refinements
+      while(refine_local != std::vector<unsigned int>(refine_local.size(), 0))
+      {
+        for(size_t material_id = 0; material_id < refine_local.size(); material_id++)
+        {
+          if(refine_local[material_id] > 0)
+          {
+            refine_local[material_id]--;
+          }
+        }
+        lambda_create_coarse_levels(0, refine_local);
```

Comment:
I don't understand the logic here. Could you explain this?

Reply:
We use the `do_create_triangulation` repeatedly to create the coarse triangulations needed. First we "undo" the global refinements, i.e., we reduce the global refinement by one each time until there is no global refinement and save the triangulation to the `coarse_triangulations` vector. Afterwards we "undo" the `local_refinements` by checking if the `refine_local` vector is all zeros, i.e., there are no local refinements left. If there are local refinements, we reduce each entity by one,build the triangulation, save it and continue reducing until there is none.

Reply:
Makes sense. I have suggested a way to simplify the code a bit.